### PR TITLE
partitions: add utility to make partition compatible filepaths

### PR DESCRIPTION
**Pull request by mr-cal**
- [x] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?

-----

Adds a utility function that takes a filepath and returns a partition-compatible filepath.

(CRAFT-1845)

### CODE REVIEW

File: craft_parts/utils/partition_utils.py [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+
+    :returns: A filepath that is compatible with the partitions feature.
+    """
+    if not Features().enable_partitions:
+        return filepath
+
+    str_path = str(filepath)
+
+    # Anything that starts with a glob should be assumed to remain that same glob.
+    if str_path.startswith("*"):
```
Comment (lengau) [inline craft_parts/utils/partition_utils.py:56]: @mr-cal I changed this a bit as I never got a satisfactory regex that met the additional requirements. Please review and lmk whether this is an acceptable alternative.

File: craft_parts/utils/partition_utils.py [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+        return filepath
```
Comment (syu-w) [inline craft_parts/utils/partition_utils.py:56]: File path could start with `*/xxx`? What's this used for?

File: tests/unit/utils/test_partition_utils.py [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+@pytest.mark.parametrize(
+    ("full_path", "expected"),
+    [
+        ("some/path", False),
+        ("(default)", True),
+        ("(default)/", True),
+        ("(part)/some/path", True),
+        ("(nota)partition", False),
+    ],
+)
```
Comment (syu-w) [inline tests/unit/utils/test_partition_utils.py:59]: `(not/a)partition`
`(not/a)/partition`
`(NOTA)/partition`
`(not1)/partition`

We really not allowed use number and capital case?

Comment (lengau) [reply tests/unit/utils/test_partition_utils.py:59]: Correct — we had a conversation over it at https://docs.google.com/document/d/1LeJvrUJAgyoGtbNtjQH6HTFnRMaWxO2Swq_BbvyrhP0/edit?disco=AAAAyZyrkx4

Thanks for the additional test cases! Added!

Comment (syu-w) [reply tests/unit/utils/test_partition_utils.py:59]: OK then, I think upper case and auto upper case, with number, maybe more better for differentiating partitions. Given that this is a historical legacy that's all for now.

File: craft_parts/utils/partition_utils.py (hunk 2 above)
Comment (lengau) [reply craft_parts/utils/partition_utils.py:56]: Some of the defaults are strings that are just `"*"`, used for overlay/stage/prime fields. I'd want to double-check with others, but it's feasible that someone might want to prime a file in all partitions.

Comment (mr-cal) [reply craft_parts/utils/partition_utils.py:56]: If I understand correctly, the question is if we should glob partitions (`*/xxx` -> `(default)/xxx, (kernel)/xxx`) or if it should apply the glob inside the default partition (`*/xxx` -> `(default)/*/xxx`).

Both options are fine, but I lean towards the second option (where the glob is applied inside the default partition).  

Two reasons:
1. If filepaths beginning with a glob behave differently, then we'll need some extra documentation (I could see this being confusing).
2. If a user wants the same file in all partitions, they can explicitly list the partitions (`stage: ["(default)/xxx", "(kernel)/xxx"]`)

Perhaps @flotter has an opinion here?

File: craft_parts/utils/partition_utils.py (hunk 1 above)
Comment (mr-cal) [reply craft_parts/utils/partition_utils.py:55]: Yeah, looking good!  I added comments elsewhere so I'll mark this as resolved.

File: tests/unit/utils/test_partition_utils.py
[the remark this answers is not in the document]
Comment (lengau) [reply tests/unit/utils/test_partition_utils.py:1]: This will happen in tests for CRAFT-1846

**Review by cmatsuoka — APPROVED** [review]
Looks good!

**Pull request merged**